[Zayd Mahmud] ip

ZaydM18 opened the pull request:
### Monty -- The Obsessive Taskmaster 🤖

Monty, when he's not hounding on you for not adding to his list, is instead tending to it and curating it instead! Some of the things he can do for you include:

- Making a to-do list
- Tracking deadlines for important tasks
- Taking note of any events you may have planned and when they'll be
- Telling you if you have anything due on a certain day
- Looking through your list to search for a particular word or phrase

For those of you who are interested in **programming**, you should know that Monty likes to maintain himself in lists too! That's right, he's compartmentalised his internal mechanisms quite well. Take a look:

```java

import monty.parser.Parser;
import monty.storage.Storage;
import monty.task.Task;
import monty.ui.Ui;
import monty.exception.MontyException;

public class Monty {
    private final Ui ui;
    private final ArrayList<Task> tasks;

```

> [!NOTE]
> Monty's design adhere's to _object-oriented programming_ principles, which is a paradigm that views the world as a network of interacting objects. Each object has a state and a behaviour, but is never too complex! You can learn more about it [here.] (https://nus-cs2103-ay2425s2.github.io/website/schedule/week1/topics.html#W1-1)

Here are some working commands that can be used with Monty:

1. 'list'
2. 'todo'
3. 'deadline' object '/by'
4. 'event' meeting '/from' start '/to' finish. 
5. 'find' 
6. 'bye'

Some features Monty is expected to have by the end are:

- [x] Text UI
- [x] Search Function
- [ ] Graphical UI
- [ ] Interactive Elements
- [ ] Conversational Ability

Commit:
Add Gradle support

ZaydM18 committed:
Level-0

ZaydM18 committed:
Level-1

ZaydM18 committed:
Level-2

ZaydM18 committed:
Level-3

ZaydM18 committed:
Level-4

ZaydM18 committed:
Automated Text Testing Added

ZaydM18 committed:
Level-4

ZaydM18 committed:
Revert "Level-4"

This reverts commit a6aa64155aa1cbc567d6f7f8ddf6c70b86517cfd.

ZaydM18 committed:
Level-5

ZaydM18 committed:
Level-6

ZaydM18 committed:
Add a Storage class to handle functionality of saving and loading task lists.

Modify the Task class to include an abstract method to act as a toString for the file and to include a protected variable that can modify the completion status.

Modify the Monty class to accommodate.

ZaydM18 committed:
no message

ZaydM18 committed:
Add date feature to convert deadlines into a more readable format.

Add function to be able to retrieve the events or deadlines due/relevant on a specific date.

LuBolin commented on src/main/java/Storage.java:
```diff
@@ -0,0 +1,69 @@
+import java.io.*;
```
Should this import be explicit, instead of using .*?

ZaydM18 replied:
You're right. It was supposed to be explicit, but I forgot to remove the wildcard after testing.

LuBolin commented on src/main/java/Monty.java:
```diff
@@ -0,0 +1,117 @@
+                        break;
+
+                    case "delete":
+                        if (argument.isEmpty()) throw new MontyException(" Your task number is out of range! How am I to keep track of my-- I mean, your -- beautiful list!");
```
Maybe we can wrap the throw in braces?

LuBolin commented on src/main/java/Event.java:
```diff
@@ -0,0 +1,35 @@
+
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + from.format(OUTPUT_FORMAT) + " to: " + to.format(OUTPUT_FORMAT) + ")";
```
Perhaps a line break would help with readability?

ZaydM18 replied:
Thanks for the suggestion. I wasn't entirely sure on where to place the separators for a line like this so I just left it as a single one.

LuBolin commented on src/main/java/Monty.java:
```diff
@@ -0,0 +1,117 @@
+                        Storage.saveTasks(tasks);
+                        break;
+
+                    default:
```
Good that you include a default clause in your switch case 👍

LuBolin reviewed (commented):
Mostly stuck with the conventions, good job!
The only thing that "needs" to be changed seems to be the .* import, and potentially having braces around one-liner if statements.
Good job!

ZaydM18 committed:
Add Ui class to handle the functionality of the chatbot interface.

Add Parser class to handle the command execution of the chatbot.

Refactor Monty to work using new classes to better reflect the Single Responsibility Principle.

ZaydM18 committed:
Organise code into packages

ZaydM18 committed:
Create packages

ZaydM18 committed:
Merge commit '68c58c12c1d631736a4b25500248a2956171496f'

ZaydM18 committed:
Test Gradle

ZaydM18 committed:
Add JUnit testing support.

Add appropriate test classes to test modular and cohesive functionality.

Modified Parser class to better handle exceptions.

ZaydM18 committed:
Add JavaDoc comments to the code

ZaydM18 committed:
Reformat to coding standard, switch-case alignment not fixing

ZaydM18 committed:
Merge branch 'branch-A-CodingStandard' into branch-A-JavaDoc

# Conflicts:
#	src/main/java/monty/Monty.java
#	src/main/java/monty/storage/Storage.java

ZaydM18 committed:
Attempt to resolve merge conflict

ZaydM18 committed:
Add 'find' feature to Parser class.

Edit Ui class to include message for the 'find' feature.

ZaydM18 committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

# Conflicts:
#	src/main/java/monty/parser/Parser.java

ZaydM18 committed:
Add method to retrieve description in Task

ZaydM18 committed:
Add more comments

ZaydM18 committed:
Merge branch 'branch-A-JavaDoc' into branch-Level-9

ZaydM18 committed:
Add filler to use tag.

Follow git standard (to the best of my knowledge).

ZaydM18 committed:
Fix issue with JavaDoc comment

ZaydM18 committed:
Merge branch 'branch-Level-9'

VikramGoyal23 commented on src/main/java/monty/Monty.java:
```diff
@@ -0,0 +1,65 @@
+import monty.task.Task;
+import monty.ui.Ui;
+import monty.exception.MontyException;
+import java.util.ArrayList;
```
A line break between these imports would help improve readability since they belong to different packages.

VikramGoyal23 commented on src/main/java/monty/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+
+import monty.exception.MontyException;
+import monty.storage.Storage;
+import monty.task.*;
```
Wildcard imports are not allowed.

VikramGoyal23 commented on src/main/java/monty/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+import monty.storage.Storage;
+import monty.task.*;
+import monty.ui.Ui;
+import java.time.LocalDate;
```
A line break between these imports would help improve readability since they belong to different packages.

VikramGoyal23 commented on src/main/java/monty/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+        String argument = words.length > 1 ? words[1].trim() : "";
+
+        switch (command) {
+            case "bye": {
```
Coding standard dictates that `switch` statements should have no indentation for their `case` clauses, outside of Lambda-style `switch` statements.

VikramGoyal23 commented on src/main/java/monty/parser/Parser.java:
```diff
@@ -0,0 +1,192 @@
+
+        return index;
+    }
+
+    private static void processDateCommand(String argument, ArrayList<Task> tasks, Ui ui) throws MontyException {
```
Good job putting header comments everywhere, but in my opinion, this private method is non-trivial enough to also warrant one.

VikramGoyal23 commented on src/main/java/monty/storage/Storage.java:
```diff
@@ -0,0 +1,111 @@
+
+        try {
+            Task task;
+            switch (type) {
+                case "T":
```
`switch` statements should have no indentation for their `case` clauses.

VikramGoyal23 commented on src/test/java/monty/MontyTest.java:
```diff
@@ -0,0 +1,58 @@
+package monty;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import java.io.ByteArrayInputStream;
```
A line break between these imports would help improve readability since they belong to different packages.

VikramGoyal23 commented on src/test/java/monty/parser/ParserTest.java:
```diff
@@ -0,0 +1,45 @@
+
+import monty.exception.MontyException;
+import monty.task.Task;
+import monty.ui.Ui;
+import org.junit.jupiter.api.BeforeEach;
```
A line break between these imports would be good.

VikramGoyal23 commented on src/test/java/monty/storage/StorageTest.java:
```diff
@@ -0,0 +1,21 @@
+package monty.storage;
+
+import monty.task.Task;
+import monty.task.ToDo;
+import org.junit.jupiter.api.Test;
+import java.util.ArrayList;
+import static org.junit.jupiter.api.Assertions.*;
```
A line break between these imports would be good.

VikramGoyal23 replied:
Wildcard imports are not allowed.

VikramGoyal23 replied:
Reorder the org.junit.jupiter imports to be together to follow the coding standard.

VikramGoyal23 commented on src/test/java/monty/task/DeadlineTest.java:
```diff
@@ -0,0 +1,19 @@
+package monty.task;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Wildcard imports are not allowed.

VikramGoyal23 commented on src/test/java/monty/task/EventTest.java:
```diff
@@ -0,0 +1,27 @@
+package monty.task;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Wildcard imports are not allowed.

VikramGoyal23 commented on src/test/java/monty/task/TaskTest.java:
```diff
@@ -0,0 +1,30 @@
+package monty.task;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Wildcard import.

VikramGoyal23 commented on src/test/java/monty/task/ToDoTest.java:
```diff
@@ -0,0 +1,19 @@
+package monty.task;
+
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
```
Wildcard import.

VikramGoyal23 commented on src/test/java/monty/ui/UiTest.java:
```diff
@@ -0,0 +1,61 @@
+package monty.ui;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import java.io.ByteArrayOutputStream;
```
Line break here would be good.

VikramGoyal23 commented on src/test/java/monty/ui/UiTest.java:
```diff
@@ -0,0 +1,61 @@
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Wildcard import.

VikramGoyal23 reviewed (commented):
LGTM, the only nitpicks I could find were in import statement formatting and wildcard imports, with some incorrect `switch` statement indentation here and there. I went ahead and marked all instances I could find.

Overall, I would say you've done a great job.

ZaydM18 committed:
Add GUI.

Add VarArg functionality.

ZaydM18 committed:
Add GUI and VarArgs

ZaydM18 committed:
Add assertions to Task, Storage and MainWindow classes

ZaydM18 committed:
Structure the code to better fit the guidelines

ZaydM18 committed:
Refactor code to better reflect Java Coding Standards

To better reflect the course's coding standards, I had to make changes
to my IDE's settings, especially to handle switch statements and method
braces.

Line breaks were more complicated to handle and required case-by-case
oversight. The coding standard will need to be maintained for future
commits.

ZaydM18 committed:
Merge pull request #2 from ZaydM18/branch-A-CodeQuality

Refactor code to better reflect Java Coding Standards

ZaydM18 committed:
Add sorting function for different task types

Added a new feature to Monty that allows users to sort their task list.

ToDo tasks can be sorted alphabetically, while deadlines can be sorted
chronologically. In the case of events, the current functionality is to
have them automatically sorted by their start dates, but other options
are being considered.

ZaydM18 committed:
Resolve merge conflicts

Adding the BCD extension before pulling my previous PR changes to the
origin repository led to some merge conflicts.

Some functionality has been altered, but a rewrite will likely be
necessary in future commits.

ZaydM18 committed:
Change processSortCommand behaviour

Previously, when sorting commands using todo, deadline and events,
clusters of the different event types would be sorted correctly but they
would not be sorted against each other.

Under the current implementation, the sort command processes tasks in
the order: todo --> deadline --> event (by start time). Therefore, the
issue of only sorting clusters of each task type has been resolved.

ZaydM18 committed:
Make aesthetic improvements to the GUI

The profile pictures were made larger. They are kept in square shapes to
match the general aesthetic being considered.

The aesthetic elements of the UI -- the background, text bar, font and
dialog boxes were changed to look more visually interesting, with
inspiration being derived from the game Final Fantasy VII.

ZaydM18 committed:
Remove unnecessary print statement

ZaydM18 committed:
Add additional functions and tune Monty's behaviour

Added a 'clear' as well as a 'confirm' function to enable the user to
clear the task list in a single command after asking for confirmation.
The delete command has also been modified to request a confirmation
beforehand.

Monty's response messages have been tweaked to add more personality to
the bot for a more fun and engaging user experience.

Some extra, unused pieces of code and some old JavaDoc comments have
been removed to make the code look cleaner.

ZaydM18 committed:
Try to resolve issues with gradle and git commit history

ZaydM18 committed:
Add user guide and clean up gradle issues

Comprehensive user guide has been updated, and issues involving gradle
appear to have been resolved.

ZaydM18 committed:
Update README.md

ZaydM18 committed:
Release Monty

Made some small adjustments to the documentation to match Monty's
behaviour more accurately.

Updated Ui.png according to changes made.

ZaydM18 committed:
Add files via upload

ZaydM18 committed:
Add 'close window' functionality to 'bye' command

ZaydM18 committed:
Amend Ui and README issues

ZaydM18 committed:
Add help command

The help command was added, which links to the product site where the
different commands are listed and explained.
